hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/ClientMessageDecoder.java

```diff
@@ -90,13 +90,14 @@ public HandlerStatus onRead() {
                     throw new IllegalStateException(
                             "Fragmented client messages are not allowed before the client is authenticated.");
                 } else {
-                    ClientMessage.ForwardFrameIterator frameIterator = activeReader.getClientMessage().frameIterator();
+                    ClientMessage message = activeReader.getClientMessage();
+                    ClientMessage.ForwardFrameIterator frameIterator = message.frameIterator();
                     //ignore the fragmentationFrame
                     frameIterator.next();
                     ClientMessage.Frame startFrame = frameIterator.next();
                     long fragmentationId = Bits.readLongL(firstFrame.content, FRAGMENTATION_ID_OFFSET);
                     if (ClientMessage.isFlagSet(flags, BEGIN_FRAGMENT_FLAG)) {
-                        builderBySessionIdMap.put(fragmentationId, ClientMessage.createForDecode(startFrame));
+                        builderBySessionIdMap.put(fragmentationId, new ClientMessage(startFrame, message.getEndFrame()));
```

Comment:
Good finding. I think we can make further cleanups. Instead of creating a new message, we put the message directly to `builderBySessionIdMap` . We just need to `dropFragmentationFrame` . 

```
                   ClientMessage message = activeReader.getClientMessage();
                    message.dropFragmentationFrame();
                    long fragmentationId = Bits.readLongL(firstFrame.content, FRAGMENTATION_ID_OFFSET);
                    if (ClientMessage.isFlagSet(flags, BEGIN_FRAGMENT_FLAG)) {
                        builderBySessionIdMap.put(fragmentationId, message);
```
We have a new method on ClientMessage:

```
  public void dropFragmentationFrame() {
        startFrame = startFrame.next;
    }
```

and merge method should change accordingly. Dropping the fragmentation logic was there, and feeling wrong to begin with. 

```
 public void merge(ClientMessage fragment) {
        endFrame.next = fragment.startFrame;
        endFrame = fragment.endFrame;
    }
```

@mdumandag How does that look ?

Reply:
Thanks for the suggestion. That version looks better.